### PR TITLE
Added filtering system to prepare_metric_for_sending.

### PR DESCRIPTION
**Pull request by gir489returns**

### CODE REVIEW

**Comment by gir489returns** [comment]
image

**Review by Rxann — COMMENTED** [review]
I have noticed some metrics usually come in waves so this might spam the notifications, but this is a debug feature so that doesn't matter.

**Comment by gir489returns** [comment]
> I have noticed some metrics usually come in waves so this might spam the notifications, but this is a debug feature so that doesn't matter.

image

File: src/hooks/info/prepare_metric_for_sending.cpp [hunk 1]
```diff
@@ -85,30 +87,44 @@ namespace big
...
 				return g_hooking->get_original<prepare_metric_for_sending>()(serializer, unk, time, metric);
 			}
 			return false;
 		}
-
-		if (!is_bad_metric && g.debug.logs.metric_logs)
+		else
 		{
-			LOG(INFO) << "METRIC: " << metric->get_name() << "; DATA: " << yim_serializer.get_string();
+			if (g.debug.logs.metric_logs == 1)
```
Comment (Yimura) [inline src/hooks/info/prepare_metric_for_sending.cpp]: else if?

File: src/hooks/info/prepare_metric_for_sending.cpp [hunk 2]
```diff
@@ -85,30 +87,44 @@ namespace big
...
-			if(strcmp(metric->get_name(), "MM") == 0)
+			if ((is_filtered_bad_metrics && g.debug.logs.metric_logs) || is_warn_bad_metrics)
+			{
+				LOG(WARNING) << "BAD METRIC: " << metric_name << "; DATA: " << yim_serializer.get_string();
+			}
+			if (is_warn_bad_metrics)
+			{
+				g_notification_service.push_warning("METRIC"_T.data(),
+				    std::format("{} {}", "METRIC_WARNING_MESSAGE"_T, metric_name).c_str());
+			}
```
Comment (Yimura) [inline src/hooks/info/prepare_metric_for_sending.cpp]: Maybe notifications of these should be made optional.

**Review by Yimura — CHANGES_REQUESTED** [review]

File: src/core/settings.hpp [hunk 3]
```diff
@@ -170,7 +170,9 @@ namespace big
 			pair out_of_allowed_range_sync_type{};
 			pair invalid_sync{};
 
-			NLOHMANN_DEFINE_TYPE_INTRUSIVE(notifications, gta_thread_kill, gta_thread_start, network_player_mgr_init, network_player_mgr_shutdown, player_join, player_leave, send_net_info_to_lobby, transaction_rate_limit, mismatch_sync_type, out_of_allowed_range_sync_type, invalid_sync)
+			bool warn_metric = true;
```
Comment (Yimura) [inline src/core/settings.hpp]: Idk, maybe the default should be false. I bet a bunch of people will whine about these popping up suddenly.

**Review by Yimura — CHANGES_REQUESTED** [review]

File: src/hooks/info/prepare_metric_for_sending.cpp [hunk 4]
```diff
@@ -87,7 +87,7 @@ namespace big
 	bool hooks::prepare_metric_for_sending(rage::json_serializer* serializer, int unk, int time, rage::rlMetric* metric)
 	{
 		char metric_json_buffer [256] {};
-		rage::json_serializer yim_serializer(metric_json_buffer, sizeof(metric_json_buffer)); //I serialized your mom's JSON buffer.
+		rage::json_serializer yim_serializer(metric_json_buffer, sizeof(metric_json_buffer));
```
Comment (gir489returns) [inline src/hooks/info/prepare_metric_for_sending.cpp:90]: image

**Review by Yimura — APPROVED** [review]

**Pull request merged**